util/types/src/utilities/difficulty.rs

```diff
@@ -1,58 +1,226 @@
-
-/// f(x) = 2^256 / x
-pub fn target_to_difficulty(boundary: &Byte32) -> U256 {
-    let d = U256::from_big_endian(boundary.as_slice()).expect("convert from Byte32 to U256");
-    if d.le(&ONE) {
+// ONE << 256
+const HSPACE: U512 = u512!("0x10000000000000000000000000000000000000000000000000000000000000000");
+
+fn target_to_difficulty(target: &U256) -> U256 {
+    if target == &ONE {
         U256::max_value()
     } else {
-        (&T_NUMBER / d) << 1u8
+        let (target, _): (U512, bool) = target.convert_into();
+        (HSPACE / target).convert_into().0
     }
 }
 
-pub fn difficulty_to_target(difficulty: &U256) -> Byte32 {
-    if difficulty.le(&ONE) {
-        Byte32::max_value()
+fn difficulty_to_target(difficulty: &U256) -> U256 {
+    if difficulty == &ONE {
+        U256::max_value()
     } else {
-        let boundary = (&T_NUMBER / difficulty) << 1u8;
-        let mut inner = vec![0u8; 32];
-        boundary
-            .into_big_endian(&mut inner)
-            .expect("convert from U256 to Byte32");
-        Byte32::new_unchecked(inner.into())
+        let (difficulty, _): (U512, bool) = difficulty.convert_into();
+        (HSPACE / difficulty).convert_into().0
     }
 }
+/**
+* the original nBits implementation inherits properties from a signed data class,
+* allowing the target threshold to be negative if the high bit of the significand is set.
+* This is useless—the header hash is treated as an unsigned number,
+* so it can never be equal to or lower than a negative target threshold.
+*
+*
+* The "compact" format is a representation of a whole
+* number N using an unsigned 32bit number similar to a
+* floating point format.
+* The most significant 8 bits are the unsigned exponent of base 256.
+* This exponent can be thought of as "number of bytes of N".
+* The lower 24 bits are the mantissa.
+* N = mantissa * 256^(exponent-3)
+*/
+fn get_low64(target: &U256) -> u64 {
+    target.0[0]
+}
+
+pub fn target_to_compact(target: U256) -> u32 {
+    let bits = 256 - target.leading_zeros();
+    let exponent = u64::from((bits + 7) / 8);
+    let mut compact = if exponent <= 3 {
+        get_low64(&target) << (8 * (3 - exponent))
+    } else {
+        get_low64(&(target >> (8 * (exponent - 3))))
+    };
+
+    compact |= exponent << 24;
+    compact as u32
+}
+
+pub fn compact_to_target(compact: u32) -> (U256, bool) {
+    let exponent = compact >> 24;
+    let mut mantissa = U256::from(compact & 0x00ff_ffff);
+
+    let mut ret;
+    if exponent <= 3 {
+        mantissa >>= 8 * (3 - exponent);
+        ret = mantissa.clone();
+    } else {
+        ret = mantissa.clone();
+        ret <<= 8 * (exponent - 3);
+    }
+
+    let overflow = !mantissa.is_zero()
+        && ((exponent > 34) // > (256 ** 31)
+            || (mantissa > U256::from(0xffu32) && exponent > 33)
+            || (mantissa > U256::from(0xffffu32) && exponent > 32));
+    (ret, overflow)
+}
```

Comment:
I am not sure whether we should add some validity check for the compact in `compact_to_target`, or we have already checked it at somewhere.

```rust
    fn test_compact_to_target2() {
        let c1 = 0x21000001;
        let (x, overflow) = compact_to_target(c1);
        println!("{:#x}", x);
        assert_eq!(overflow, false, "should not overflow");

        let c = target_to_compact(x);
        println!("{:#x}", c);
        assert_eq!(c, c1);
    }
```

This test case will fail because even the compact is not overflowed, this compact is not a rightful number which could convert from the target.

Reply:
target is verified by consensus rule,  
so there is no correctness issue here.

this question depends on how you define overflow.
if consider compact_to_target individually, treat compact as any 32-bit integer, so the limit is 256U::max
if consider precision and target_to_compact, 0x21.. should not appear.